ssh-server/main.go

```diff
@@ -111,40 +116,78 @@ func serviceChannel(newChannel ssh.NewChannel) error {
 	if err != nil {
 		return fmt.Errorf("accepting channel: %v", err)
 	}
-	defer channel.Close()
 
-	// Start the command with a pseudo-terminal.
+	allow := map[string]bool{"shell": true, "exec": true, "pty-req": true, "window-change": true}
+	resize := make(chan *pty.Winsize, 1)
+	done := make(chan struct{})
+	for {
+		select {
+		case <-done:
+			return nil
+		case req, ok := <-requests:
+			if !ok {
+				return nil
+			}
+			if req.WantReply {
+				req.Reply(allow[req.Type], nil)
+			}
+			switch req.Type {
+			case "pty-req":
+				go func() {
+					newPTY(channel, resize)
+					close(done)
+				}()
+
+				skip := binary.BigEndian.Uint32(req.Payload)
+				size, err := windowSize(req.Payload[4+skip:])
+				if err != nil {
+					log.Printf("Failed to get window size: %v", err)
+					continue
+				}
+				resize <- size
+			case "window-change":
+				size, err := windowSize(req.Payload)
+				if err != nil {
+					log.Printf("Failed to get window size: %v", err)
+					continue
+				}
+				resize <- size
+			}
+		}
+	}
+}
+
+func newPTY(channel ssh.Channel, resize chan *pty.Winsize) {
+	//Start the command with a pseudo-terminal.
 	shell, err := pty.Start(exec.Command("bash"))
 	if err != nil {
-		return fmt.Errorf("starting pty: %v", err)
+		log.Printf("starting pty: %v", err)
 	}
 	defer shell.Close()
 
-	go func(in <-chan *ssh.Request) {
-		for req := range in {
-			// Tell the client we accept pty and shell commands
-			req.Reply(req.Type == "pty-req" || req.Type == "shell", nil)
-
-			if req.Type == "pty-req" {
-				skip := binary.BigEndian.Uint32(req.Payload)
-				resize(shell, req.Payload[4+skip:])
-			}
-			if req.Type == "window-change" {
-				resize(shell, req.Payload)
-			}
+	go func() {
+		for size := range resize {
+			pty.Setsize(shell, size)
 		}
-	}(requests)
+	}()
 
 	go io.Copy(shell, channel)
 	io.Copy(channel, shell)
-	return nil
+
+	status := struct{ Status uint32 }{uint32(0)}
+	if _, err := channel.SendRequest("exit-status", false, ssh.Marshal(&status)); err != nil {
+		log.Printf("Failed to send exit request: %v", err)
+	}
+	channel.Close()
```

Comment:
Is there a reason to only close this channel here?  Why not just defer channel.Close() on line 118 after Accept?

Reply:
It's important to close this channel so the go routine will finish and I wanted to make it obvious but defering it will close it even in error cases which is what we want